File: apps/web/lib/mesocycle.ts
/** Client-side mesocycle plan (mirrors garmin_coaching/mesocycle.py). */

export const MESOCYCLE_ANCHOR = "2026-09-21";

const PHASE_WEEKLY_BASE: Record<string, Record<string, number>> = {
  R0: { run: 60, bike: 30, swim: 0, walk: 120 },
  R1: { run: 200, bike: 150, swim: 0, walk: 45 },
  R2: { run: 280, bike: 60, swim: 0, walk: 30 },
  R3: { run: 320, bike: 40, swim: 0, walk: 30 },
  R4: { run: 180, bike: 0, swim: 0, walk: 60 },
  R5: { run: 90, bike: 280, swim: 0, walk: 30 },
  R6: { run: 60, bike: 320, swim: 0, walk: 30 },
  R7: { run: 150, bike: 200, swim: 120, walk: 30 },
};

const PHASES: { code: string; start: string; end: string; name: string }[] = [
  { code: "R0", start: "2026-09-20", end: "2026-10-05", name: "Recuperacion" },
  { code: "R1", start: "2026-10-06", end: "2026-12-31", name: "Base aerobica" },
  { code: "R2", start: "2027-01-01", end: "2027-03-31", name: "Construccion" },
  { code: "R3", start: "2027-04-01", end: "2027-07-24", name: "Especifico 21K" },
  { code: "R4", start: "2027-07-25", end: "2027-07-25", name: "Carrera mmB" },
  { code: "R5", start: "2027-08-01", end: "2027-08-31", name: "Transicion ciclismo" },
  { code: "R6", start: "2027-09-01", end: "2027-09-13", name: "Reto Letras" },
  { code: "R7", start: "2027-10-01", end: "2027-11-28", name: "Especifico 70.3" },
];

const LOAD_WEEK_SCALE = [0.85, 1.0, 1.1];
const RECOVERY_WEEK_SCALE = 0.65;

export function mondayOf(isoDate: string): string {
  const d = new Date(isoDate + "T12:00:00");
  const day = d.getDay();
  const diff = day === 0 ? -6 : 1 - day;
  d.setDate(d.getDate() + diff);
  return d.toISOString().slice(0, 10);
}

function activePhaseCode(isoDate: string): string {
  for (const p of PHASES) {
    if (isoDate >= p.start && isoDate <= p.end) return p.code;
  }
  if (isoDate < PHASES[0].start) return PHASES[0].code;
  return PHASES[PHASES.length - 1].code;
}

export function weekMeta(weekStart: string) {
  const ws = mondayOf(weekStart);
  const anchor = new Date(MESOCYCLE_ANCHOR + "T12:00:00");
  const cur = new Date(ws + "T12:00:00");
  const days = Math.floor((cur.getTime() - anchor.getTime()) / 86400000);
  let blockIndex = 0;
  let weekInBlock = 0;
  if (days >= 0) {
    const weekIndex = Math.floor(days / 7);
    blockIndex = Math.floor(weekIndex / 4);
    weekInBlock = weekIndex % 4;
  }
  const kind = weekInBlock === 3 ? "recovery" : "load";
  return { week_start: ws, block_index: blockIndex, week_in_block: weekInBlock, kind, phase_code: activePhaseCode(ws) };
}

export function plannedWeeklySports(weekStart: string) {
  const meta = weekMeta(weekStart);
  const base = { ...(PHASE_WEEKLY_BASE[meta.phase_code] ?? PHASE_WEEKLY_BASE.R0) };
  const scale = meta.week_in_block === 3 ? RECOVERY_WEEK_SCALE : LOAD_WEEK_SCALE[meta.week_in_block];
  return {
    run: Math.round(base.run * scale * 10) / 10,
    bike: Math.round(base.bike * scale * 10) / 10,
    swim: Math.round(base.swim * scale * 10) / 10,
    walk: Math.round(base.walk * scale * 10) / 10,
  };
}

export type WeeklySportRow = {
  week_start: string;
  block_index: number | null;
  week_in_block: number | null;
  kind: string | null;
  phase_code: string | null;
  planned_run_min: number | null;
  planned_bike_min: number | null;
  planned_swim_min: number | null;
  planned_walk_min: number | null;
  planned_total_min: number | null;
  executed_run_min: number | null;
  executed_bike_min: number | null;
  executed_swim_min: number | null;
  executed_walk_min: number | null;
  executed_total_min: number | null;
};

export function mergePlannedExecuted(row: WeeklySportRow): WeeklySportRow {
  const planned = plannedWeeklySports(row.week_start);
  const plannedTotal = planned.run + planned.bike + planned.swim + planned.walk;
  const meta = weekMeta(row.week_start);
  return {
    ...row,
    ...meta,
    planned_run_min: row.planned_run_min ?? planned.run,
    planned_bike_min: row.planned_bike_min ?? planned.bike,
    planned_swim_min: row.planned_swim_min ?? planned.swim,
    planned_walk_min: row.planned_walk_min ?? planned.walk,
    planned_total_min: row.planned_total_min ?? plannedTotal,
    executed_run_min: row.executed_run_min ?? 0,
    executed_bike_min: row.executed_bike_min ?? 0,
    executed_swim_min: row.executed_swim_min ?? 0,
    executed_walk_min: row.executed_walk_min ?? 0,
    executed_total_min: row.executed_total_min ?? 0,
  };
}

File: garmin_coaching/mesocycle.py
"""3+1 mesocycles: planned weekly volume by sport (Monday always recovery)."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, timedelta
from typing import Any

from garmin_coaching.activity_load import monday_of
from garmin_coaching.macrocycle import MacrocyclePhase, active_phase

# First Monday on or after macrocycle R0 start (2026-09-21).
MESOCYCLE_ANCHOR = date(2026, 9, 21)

# Base weekly minutes at load week 2 (sport totals; Monday recovery baked into daily split).
_PHASE_WEEKLY_BASE: dict[str, dict[str, int]] = {
    "R0": {"run": 60, "bike": 30, "swim": 0, "walk": 120},
    "R1": {"run": 200, "bike": 150, "swim": 0, "walk": 45},
    "R2": {"run": 280, "bike": 60, "swim": 0, "walk": 30},
    "R3": {"run": 320, "bike": 40, "swim": 0, "walk": 30},
    "R4": {"run": 180, "bike": 0, "swim": 0, "walk": 60},
    "R5": {"run": 90, "bike": 280, "swim": 0, "walk": 30},
    "R6": {"run": 60, "bike": 320, "swim": 0, "walk": 30},
    "R7": {"run": 150, "bike": 200, "swim": 120, "walk": 30},
}

_LOAD_WEEK_SCALE = (0.85, 1.0, 1.1)
_RECOVERY_WEEK_SCALE = 0.65
_MONDAY_WALK_MIN = 30


@dataclass(frozen=True)
class MesocycleWeek:
    week_start: date
    block_index: int
    week_in_block: int
    kind: str
    phase_code: str


def week_meta(week_start: date) -> MesocycleWeek:
    ws = monday_of(week_start)
    days = (ws - MESOCYCLE_ANCHOR).days
    if days < 0:
        block_index, week_in_block = 0, 0
    else:
        week_index = days // 7
        block_index = week_index // 4
        week_in_block = week_index % 4
    kind = "recovery" if week_in_block == 3 else "load"
    phase = active_phase(ws)
    return MesocycleWeek(
        week_start=ws,
        block_index=block_index,
        week_in_block=week_in_block,
        kind=kind,
        phase_code=phase.code,
    )


def _scale_for_week(week_in_block: int) -> float:
    if week_in_block == 3:
        return _RECOVERY_WEEK_SCALE
    return _LOAD_WEEK_SCALE[week_in_block]


def planned_weekly_sports(week_start: date) -> dict[str, float]:
    meta = week_meta(week_start)
    base = dict(_PHASE_WEEKLY_BASE.get(meta.phase_code, _PHASE_WEEKLY_BASE["R0"]))
    scale = _scale_for_week(meta.week_in_block)
    return {sport: round(mins * scale, 1) for sport, mins in base.items()}


def _daily_non_monday_budget(weekly: dict[str, float]) -> dict[str, float]:
    """Reserve Monday walk; spread remaining planned minutes Tue–Sun."""
    out = dict(weekly)
    walk = out.get("walk", 0)
    out["walk"] = max(0.0, walk - _MONDAY_WALK_MIN)
    days = 6
    return {k: (v / days if k != "swim" else v / max(3, days // 2)) for k, v in out.items()}


def planned_daily_sports(on_day: date) -> dict[str, float]:
    ws = monday_of(on_day)
    weekly = planned_weekly_sports(ws)
    if on_day.weekday() == 0:
        return {"run": 0.0, "bike": 0.0, "swim": 0.0, "walk": float(_MONDAY_WALK_MIN)}
    per_day = _daily_non_monday_budget(weekly)
    return {k: round(v, 1) for k, v in per_day.items()}


def plan_week(week_start: date) -> dict[str, Any]:
    ws = monday_of(week_start)
    meta = week_meta(ws)
    phase = active_phase(ws)
    planned = planned_weekly_sports(ws)
    daily: list[dict[str, Any]] = []
    for i in range(7):
        d = ws + timedelta(days=i)
        daily.append({"date": d.isoformat(), "sports": planned_daily_sports(d)})

    return {
        "week_start": ws.isoformat(),
        "block_index": meta.block_index,
        "week_in_block": meta.week_in_block,
        "week_label": f"Bloque {meta.block_index + 1} · Semana {meta.week_in_block + 1}/4",
        "kind": meta.kind,
        "phase_code": meta.phase_code,
        "phase_name": phase.name,
        "planned_run_min": planned.get("run", 0),
        "planned_bike_min": planned.get("bike", 0),
        "planned_swim_min": planned.get("swim", 0),
        "planned_walk_min": planned.get("walk", 0),
        "planned_total_min": round(sum(planned.values()), 1),
        "daily": daily,
    }


def iter_weeks(from_date: date, to_date: date) -> list[date]:
    start = monday_of(from_date)
    end = monday_of(to_date)
    out: list[date] = []
    cur = start
    while cur <= end:
        out.append(cur)
        cur += timedelta(days=7)
    return out


def phase_for_week(week_start: date) -> MacrocyclePhase:
    return active_phase(monday_of(week_start))


def mesocycle_summary_until(end: date) -> list[dict[str, Any]]:
    start = MESOCYCLE_ANCHOR
    return [plan_week(ws) for ws in iter_weeks(start, end)]
